lib/collection/src/collection/distance_matrix.rs

```diff
@@ -0,0 +1,142 @@
+use std::collections::HashSet;
+use std::time::Duration;
+
+use rand::prelude::SliceRandom;
+use segment::data_types::vectors::{
+    NamedMultiDenseVector, NamedVectorStruct, VectorStructInternal,
+};
+use segment::types::{Filter, HasIdCondition, PointIdType, ScoredPoint, WithVector};
+
+use crate::collection::Collection;
+use crate::operations::query_enum::QueryEnum;
+use crate::operations::shard_selector_internal::ShardSelectorInternal;
+use crate::operations::types::{
+    CollectionResult, CoreSearchRequest, CoreSearchRequestBatch, PointRequestInternal,
+};
+
+#[derive(Debug, Default)]
+struct DistanceMatrixResult {
+    sample_ids: Vec<PointIdType>,   // sampled point ids
+    nearest: Vec<Vec<ScoredPoint>>, // nearest points for each sampled point
+}
+
+impl Collection {
+    async fn distance_matrix(
+        &self,
+        sample_count: usize,
+        limit: usize,
+        filter: Option<&Filter>,
+        using: String,
+        timeout: Option<Duration>,
+    ) -> CollectionResult<DistanceMatrixResult> {
+        if limit == 0 || sample_count == 0 {
+            return Ok(Default::default());
+        }
+
+        let shard_selector = ShardSelectorInternal::All;
+
+        let mut sampled_point_ids = Vec::with_capacity(sample_count);
+
+        // Get local shards in random order
+        let mut local_shards = self.get_local_shards().await;
+        local_shards.shuffle(&mut rand::thread_rng());
+
+        // Sample points from local shards
+        for local_shard in self.get_local_shards().await {
+            // Break if we have enough sample points
+            if sampled_point_ids.len() >= sample_count {
+                break;
+            }
+            let guard = self.shards_holder.read().await;
+            if let Some(shard) = guard.get_shard(&local_shard) {
+                let sampled = shard
+                    .sample_filtered_points(sample_count, filter, timeout)
+                    .await?;
+                sampled_point_ids.extend(sampled);
+            }
+        }
```

Comment:
Why not use the query api with `ScoringQuery::Sample(Sample::Random)`?

This current implementation also has preference towards the first shard it chooses

Reply:
- we want to sample on the local shard only to minimize network access

- regarding the preference of the first shard, I agree, an alternative implementation is to query all shards and then merge, shuffle and truncate the results

Reply:
I decided to use the Query API instead of local sampling + retrieval.
It makes the code much simpler and less weird.

Thank you for the suggestion :+1: